### PR TITLE
Use Templating for Reverse Engineering

### PR DESCRIPTION
**Pull request by lajones**
With this PR Reverse Engineering would use templating (an adjusted-copy of the templating code from Scaffolding) to produce the generated code. And updated to use separate .cshtml files for the templates (rather than inline strings).

Also made the updates which were already made in the release branch (#1928, #1934, #1937 & #1938) and which needed equivalent changes in the template-based code in this branch plus a few minor changes to whitespace and to support API changes that have been made in dev.

### CODE REVIEW

File: src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+            var fileMetadataReference = metadataReference as IMetadataFileReference;
+            if (fileMetadataReference != null)
+            {
+                var metadata = AssemblyMetadata.CreateFromStream(File.OpenRead(fileMetadataReference.Path));
+                _references.Add(metadata.GetReference());
+                return;
+            }
+
+            throw new InvalidOperationException("Unable to create " + typeof(MetadataReference).Name + " for name " + name);
```
Comment (lajones) [inline src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs]: I'll change this to be a resource.

**Comment by lajones** [comment]
@bricelam @divega @rowanmiller This is the same as the remote branch I had set up for you to look at this. Would like to get this in as it's a pre-req for many of the other changes (or at least the other changes would be different if we were to choose not to do this).

File: src/EntityFramework.Relational.Design/ReverseEngineering/DbContextCodeGeneratorHelper.cs [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+                        "ForRelational()",
+                        string.Format(CultureInfo.InvariantCulture,
+                            "DefaultExpression({0})",
+                                CSharpUtilities.Instance.DelimitString(
+                                    propertyConfiguration.Property.Relational().DefaultExpression))));
+            }
+        }
+    }
+
+    public class EntityConfiguration
```
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/DbContextCodeGeneratorHelper.cs]: Have pulled this and the other XXXConfiguration classes out into separate .cs files in a future PR.

**Comment by lajones** [comment]
@bricelam @divega @rowanmiller - pinging...

File: src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        {
+            _references.Add(MetadataReference.CreateFromAssembly(
+                Assembly.Load(new AssemblyName("mscorlib"))));
+            _references.Add(MetadataReference.CreateFromAssembly(
+                Assembly.Load(new AssemblyName("System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089"))));
+            _references.Add(MetadataReference.CreateFromAssembly(
+                Assembly.Load(new AssemblyName("System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089"))));
+            _references.Add(MetadataReference.CreateFromAssembly(
+                Assembly.Load(new AssemblyName("Microsoft.CSharp, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"))));
+            AddReferenceFromName("EntityFramework.Relational.Design");
```
Comment (bricelam) [inline src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs]: Are these right for .NET Core?

File: src/EntityFramework.Relational.Design/Templating/Compilation/RoslynCompilationService.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                    {
+                        var formatter = new DiagnosticFormatter();
+                        var errorMessages = result.Diagnostics
+                                             .Where(IsError)
+                                             .Select(d => formatter.Format(d));
+
+                        return CompiledAssemblyResult.FromErrorMessages(errorMessages);
+                    }
+
+                    ms.Seek(0, SeekOrigin.Begin);
```
Comment (bricelam) [inline src/EntityFramework.Relational.Design/Templating/Compilation/RoslynCompilationService.cs]: I don't think this is needed. `.ToArray()` don't care.

File: src/EntityFramework.Relational.Design/Templating/RazorReverseEngineeringBase.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+        public virtual dynamic Model { get;[param: NotNull]set; }
+
+        public abstract Task ExecuteAsync();
+
+        public virtual Task ExecuteAsync(
+            CancellationToken cancellationToken)
+        {
+            // This method is here solely so we can pass the ApiConsistencyTest
+            // Async_methods_should_have_overload_with_cancellation_token_and_end_with_async_suffix.
```
Comment (bricelam) [inline src/EntityFramework.Relational.Design/Templating/RazorReverseEngineeringBase.cs]: Why not make this one abstract and remove the other one?

Comment (lajones) [inline src/EntityFramework.Relational.Design/Templating/RazorReverseEngineeringBase.cs]: The Razor engine expects the ExecuteAsync() with no args method.

File: src/EntityFramework.Relational.Design/Templating/Compilation/RoslynCompilationService.cs (hunk 4 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/Templating/Compilation/RoslynCompilationService.cs]: Don't follow?

**Comment by bricelam** [comment]
Shouldn't most of the templating code be in Commands, not Relational.Design?

**Comment by bricelam** [comment]
:shipit:

File: src/EntityFramework.Relational.Design/Templating/Compilation/RoslynCompilationService.cs (hunk 4 above)
Comment (bricelam) [inline src/EntityFramework.Relational.Design/Templating/Compilation/RoslynCompilationService.cs]: I don't see anything that requires the stream to be reset back to the beginning.

File: src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs (hunk 3 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs]: At the moment this supports net45, dnx451 and dnxcore50. I haven't updated for netcore451 yet. Let me do that as a separate PR.

File: src/EntityFramework.Relational.Design/Templating/Compilation/RoslynCompilationService.cs (hunk 4 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/Templating/Compilation/RoslynCompilationService.cs]: I see. Just didn't understand the `.ToArray()` part. Otherwise agreed.

**Comment by lajones** [comment]
I don't mind if we move the Templating part into Commands. It could be useful if something else Command-based wants to use templating in the future. And it would set the stage for localizing the actual IO into a service which I had intended to do later for testing purposes anyway. But again let me do that as a separate PR. Thanks for the review.

File: src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs (hunk 3 above)
Comment (ErikEJ) [inline src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs]: @lajones Is there some docs or blog post explaining all these new target farmework variants? (too many, IMHO)

File: src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        public virtual void AddReferenceFromName([NotNull]string name)
+        {
+            Check.NotEmpty(name, nameof(name));
+
+#if DNX451 || DNXCORE50
+            var libraryManager = _serviceProvider.GetRequiredService<ILibraryManager>();
+            var metadataReference = libraryManager.GetLibraryExport(name).MetadataReferences.Single();
```
Comment (davidfowl) [inline src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs]: A library can provide multiple metadata references. Packages do this.

File: src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs (hunk 3 above)
Comment (davidfowl) [inline src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs]: @bricelam Nope, it's not right for anything bull full CLR.

File: src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs (hunk 6 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs]: Have updated to loop over them instead.

**Comment by lajones** [comment]
Checked in with commit fd64cf8.

File: src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs (hunk 3 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs]: @ErikEJ I don't know of one but I'll try to get you an answer for this.

Comment (bricelam) [inline src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs]: @ErikEJ You may have already read it, but [ntroducing .NET Core](http://blogs.msdn.com/b/dotnet/archive/2014/12/04/introducing-net-core.aspx) has a lot of great information in it.

Comment (ErikEJ) [inline src/EntityFramework.Relational.Design/Templating/Compilation/MetadataReferencesProvider.cs]: @bricelam & @lajones  Thanks, I read that... Slowly beginning to understand
